File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root: return []
        
        res = [[]]
        queue = deque()
        queue.append((root, 0))
        
        while queue:
            cur, idx = queue.popleft()
            
            if not cur: continue
            
            if len(res) == idx:
                res.append([])
            res[idx].append(cur.val)
            queue.append((cur.left, idx+1))
            queue.append((cur.right, idx+1))

        return res

File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        res.add(new ArrayList<>());
        Queue<Node> queue = new LinkedList<>();
        queue.add(new Node(0, root));
        
        while (!queue.isEmpty()) {
            Node cur = queue.poll();
            TreeNode tree = cur.root;
            if (Objects.equals(tree, null) || Objects.equals(tree.val, null)) {
                continue;
            }
            if (res.size() == cur.idx) {
                res.add(new ArrayList<>());
            }
            List<Integer> tmp = res.get(cur.idx);
            tmp.add(tree.val);
            queue.add(new Node(cur.idx+1, tree.left));
            queue.add(new Node(cur.idx+1, tree.right));
        }
        return res;
    }
    
    static class Node {
        private int idx;
        private TreeNode root;
        
        public Node(int idx, TreeNode root) {
            this.idx = idx;
            this.root = root;
        }
    }
}